substrate/client/network/types/src/multiaddr.rs

```diff
@@ -0,0 +1,251 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use litep2p::types::multiaddr::{
+	Error as LiteP2pError, Iter as LiteP2pIter, Multiaddr as LiteP2pMultiaddr,
+	Protocol as LiteP2pProtocol,
+};
+use std::{
+	fmt::{self, Debug, Display},
+	str::FromStr,
+};
+
+mod protocol;
+pub use protocol::Protocol;
+
+// Re-export the macro under shorter name under `multiaddr`.
+pub use crate::build_multiaddr as multiaddr;
+
+/// [`Multiaddr`] type used in Substrate. Converted to libp2p's `Multiaddr`
+/// or litep2p's `Multiaddr` when passed to the corresponding network backend.
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
+pub struct Multiaddr {
```

Comment:
This version is missing `impl From<IpAddr> for Multiaddr` from libp2p's implementation, can you add it?

Reply:
Added in https://github.com/paritytech/polkadot-sdk/pull/4855.